docs/configuration.md
# Configuration

Fireactions uses YAML for configuration. Both the client and server search for a configuration file in the following locations:

1. `/etc/fireactions/config.yaml`
2. `~/.fireactions/config.yaml`
3. `./config.yaml`

## Client configuration

Example configuration for a client:

```yaml
---
fireactions_server_url: http://127.0.0.1:8080

# Name of the client. This is used to identify the client on the server.
# Defaults to the $HOSTNAME environment variable.
name: ""

# CPU and RAM overcommit ratios. These are used to calculate the resources required for a job.
cpu_overcommit_ratio: 1.0
ram_overcommit_ratio: 1.0

# Labels to apply for the client. These are used for affinity rules.
labels:
  fireactions/region: default

# The client will reconcile the state of the server with the local state every `reconcile_interval` seconds.
reconcile_interval: 5s

# Log level. Can be one of: debug, info, warn, error, fatal, panic, trace.
log_level: debug
```

## Server configuration

Example configuration for a server:

```yaml
---
# Path to the directory where the server will store its data.
data-dir: /var/lib/fireactions

# HTTP server configuration.
http:
# The address to listen on for HTTP requests.
  listen_addr: 0.0.0.0:8081

# GitHub configuration.
github:
# The GitHub webhook secret. This should be the same as the one configured in the GitHub App settings.
  webhook_secret: SECRET
# The GitHub App ID and key. This can be found in the GitHub App settings.
  app_id: 123456
  app_private_key: |
    -----BEGIN RSA PRIVATE KEY-----

# Metrics server configuration. This is used to expose Prometheus metrics on endpoint `/metrics`.
metrics:
# Whether to enable the metrics server.
  enabled: true
# The address to listen on for HTTP requests.
  address: 0.0.0.0:8082

# History (workflow runs and jobs) configuration.
history:
# The maximum age of a workflow run. Workflow runs older than this will be deleted.
  max_workflow_run_age: 24h

# List of job labels.
job_labels:
- name: fireactions-2vcpu-2gb
# List of allowed repositories. Regular expressions are supported.
  allowed_repositories:
  - ".*"
# Template for the GitHub runner name. The following variables are supported:
# - {{ .ID }}: The ID of the runner.
  runner_name_template: fireactions-2vcpu-2gb-{{ .ID }}
# Additional labels to apply to the GitHub runner. The default labels include only the job_label name.
  runner_labels:
  - ubuntu-22.04
# The GitHub runner image to use. This should be a Docker image that contains the GitHub runner and the Fireactions agent.
  runner_image: ghcr.io/hostinger/fireactions/runner:ubuntu-20.04-x64-2.310.2
# The GitHub runner image pull policy. Can be one of: Always, Never, IfNotPresent.
  runner_image_pull_policy: IfNotPresent
# The GitHub runner resources. These are used to calculate the resources required for a job.
  runner_resources:
    memory_mb: 2048
    vcpus: 2
# Affinity rules. These are used to schedule GitHub runners to specific clients. The following operators are supported:
# - NotIn: The value must not be in the list of values.
# - In: The value must be in the list of values.
  runner_affinity:
  - { key: fireactions/region, operator: In, values: [default] }
# Metadata to apply to the GitHub runner. Use with MMDS service to provide metadata to the virtual machine.
  runner_metadata:
    example1: value1
    example2: value2

# Log level. Can be one of: debug, info, warn, error, fatal, panic, trace.
log_level: debug
```


CHANGELOG.md
# Changelog

## [0.1.2](https://github.com/hostinger/fireactions/compare/fireactions-v0.1.1...fireactions-v0.1.2) (2023-12-11)


### Bug Fixes

* **images:** Add RUNNER_TOOL_CACHE directory ([e7ec617](https://github.com/hostinger/fireactions/commit/e7ec617575c1cecc53f8eeafee631e810b1eb55b))

## [0.1.1](https://github.com/hostinger/fireactions/compare/fireactions-v0.1.0...fireactions-v0.1.1) (2023-12-05)


### Bug Fixes

* **ci:** Set executable bit on fireactions binary in Dockerfile ([57915d7](https://github.com/hostinger/fireactions/commit/57915d7b8063f73564fcbc83c6ca21e049fafa06))

## [0.1.0](https://github.com/hostinger/fireactions/compare/fireactions-v0.0.1...fireactions-v0.1.0) (2023-12-05)


### Features

* Initial commit ([7fe9daf](https://github.com/hostinger/fireactions/commit/7fe9daf36e60f01edffa455b5add4806fc18ec50))


server/config.go
package server

import (
	"fmt"
	"os"
	"regexp"
	"strings"
	"text/template"
	"time"

	"github.com/hashicorp/go-multierror"
	"github.com/hostinger/fireactions"
	"github.com/samber/lo"
)

// Config is the configuration for the Server.
type Config struct {
	HTTP      *HTTPConfig       `mapstructure:"http"`
	DataDir   string            `mapstructure:"data_dir"`
	GitHub    *GitHubConfig     `mapstructure:"github"`
	Metrics   *MetricsConfig    `mapstructure:"metrics"`
	History   *HistoryConfig    `mapstructure:"history"`
	JobLabels []*JobLabelConfig `mapstructure:"job_labels"`
	LogLevel  string            `mapstructure:"log_level"`
}

// NewConfig creates a new Config with default values.
func NewConfig() *Config {
	cfg := &Config{
		HTTP:      &HTTPConfig{ListenAddress: ":8080"},
		DataDir:   "",
		GitHub:    &GitHubConfig{WebhookSecret: "", AppID: 0, AppPrivateKey: ""},
		Metrics:   &MetricsConfig{Enabled: true, Address: ":8081"},
		History:   &HistoryConfig{MaxWorkflowRunAge: time.Hour * 24 * 7},
		JobLabels: []*JobLabelConfig{},
		LogLevel:  "info",
	}

	return cfg
}

// Validate validates the configuration.
func (c *Config) Validate() error {
	var errs error

	if c.HTTP == nil {
		errs = multierror.Append(errs, fmt.Errorf("http config is required"))
	} else {
		if err := c.HTTP.Validate(); err != nil {
			errs = multierror.Append(errs, fmt.Errorf("invalid http config: %w", err))
		}
	}

	if c.DataDir == "" {
		errs = multierror.Append(errs, fmt.Errorf("data_dir is required"))
	} else {
		_, err := os.Stat(c.DataDir)
		if err != nil && os.IsNotExist(err) {
			errs = multierror.Append(errs, fmt.Errorf("data_dir does not exist: %s", c.DataDir))
		}
	}

	if c.GitHub == nil {
		errs = multierror.Append(errs, fmt.Errorf("github config is required"))
	} else {
		if err := c.GitHub.Validate(); err != nil {
			errs = multierror.Append(errs, fmt.Errorf("invalid github config: %w", err))
		}
	}

	if c.Metrics == nil {
		errs = multierror.Append(errs, fmt.Errorf("metrics config is required"))
	} else {
		if err := c.Metrics.Validate(); err != nil {
			errs = multierror.Append(errs, fmt.Errorf("invalid metrics config: %w", err))
		}
	}

	for _, label := range c.JobLabels {
		if err := label.Validate(); err != nil {
			errs = multierror.Append(errs, fmt.Errorf("invalid job_label (%s): %w", label.Name, err))
		}
	}

	switch c.LogLevel {
	case "trace", "debug", "info", "warn", "error":
	case "":
		errs = multierror.Append(errs, fmt.Errorf("log_level is required"))
	default:
		errs = multierror.Append(errs, fmt.Errorf("log_level must be one of: trace, debug, info, warn, error"))
	}

	return errs
}

// FindJobLabel finds a JobLabel by name.
func (c *Config) FindJobLabel(name string) (*JobLabelConfig, bool) {
	for _, label := range c.JobLabels {
		if label.Name != name {
			continue
		}

		return label, true
	}

	return nil, false
}

func (c *Config) FindMatchingJobLabel(labels []string) (*JobLabelConfig, bool) {
	for _, label := range labels {
		jobLabel, ok := c.FindJobLabel(label)
		if !ok {
			continue
		}

		return jobLabel, true
	}

	return nil, false
}

// HTTPConfig is the configuration for the HTTP server.
type HTTPConfig struct {
	ListenAddress string `mapstructure:"listen_addr"`
}

// Validate validates the configuration.
func (c *HTTPConfig) Validate() error {
	var errs error

	if c.ListenAddress == "" {
		errs = multierror.Append(errs, fmt.Errorf("listen_addr is required"))
	}

	return errs
}

// GitHubConfig is the configuration for the GitHub integration.
type GitHubConfig struct {
	WebhookSecret string `mapstructure:"webhook_secret"`
	AppPrivateKey string `mapstructure:"app_private_key"`
	AppID         int64  `mapstructure:"app_id"`
}

// Validate validates the configuration.
func (c *GitHubConfig) Validate() error {
	var errs error

	if c.WebhookSecret == "" {
		errs = multierror.Append(errs, fmt.Errorf("webhook_secret is required"))
	}

	if c.AppPrivateKey == "" {
		errs = multierror.Append(errs, fmt.Errorf("app_private_key is required"))
	}

	if c.AppID == 0 {
		errs = multierror.Append(errs, fmt.Errorf("app_id is required"))
	}

	return errs
}

// JobLabelConfig is the configuration for a single job label. The label defines which repositories are allowed
// to use the label and how the jobs are executed.
type JobLabelConfig struct {
	Name                  string                            `mapstructure:"name"`
	AllowedRepositories   []string                          `mapstructure:"allowed_repositories"`
	RunnerLabels          []string                          `mapstructure:"runner_labels"`
	RunnerNameTemplate    string                            `mapstructure:"runner_name_template"`
	RunnerImage           string                            `mapstructure:"runner_image"`
	RunnerImagePullPolicy fireactions.RunnerImagePullPolicy `mapstructure:"runner_image_pull_policy"`
	RunnerResources       fireactions.RunnerResources       `mapstructure:"runner_resources"`
	RunnerAffinity        []*fireactions.RunnerAffinityRule `mapstructure:"runner_affinity"`
	RunnerMetadata        map[string]interface{}            `mapstructure:"runner_metadata"`
}

// Validate validates the configuration.
func (c *JobLabelConfig) Validate() error {
	var errs error

	if c.Name == "" {
		errs = multierror.Append(errs, fmt.Errorf("name is required"))
	}

	if len(c.AllowedRepositories) == 0 {
		errs = multierror.Append(errs, fmt.Errorf("allowed_repositories is required"))
	}

	if c.RunnerNameTemplate == "" {
		errs = multierror.Append(errs, fmt.Errorf("runner_name_template is required"))
	}

	if _, err := template.New("runner_name").Parse(c.RunnerNameTemplate); err != nil {
		errs = multierror.Append(errs, fmt.Errorf("invalid runner_name_template: %w", err))
	}

	if c.RunnerImage == "" {
		errs = multierror.Append(errs, fmt.Errorf("runner_image is required"))
	}

	if c.RunnerImagePullPolicy == "" {
		errs = multierror.Append(errs, fmt.Errorf("runner_image_pull_policy is required"))
	}

	if err := c.RunnerResources.Validate(); err != nil {
		errs = multierror.Append(errs, fmt.Errorf("invalid runner_resources: %w", err))
	}

	return errs
}

// MustGetRunnerName renders the runner name from the template. It panics if the template is invalid.
func (c *JobLabelConfig) MustGetRunnerName(runnerID string) string {
	name, err := c.GetRunnerName(runnerID)
	if err != nil {
		panic(err)
	}

	return name
}

func (c *JobLabelConfig) IsAllowedRepository(repository string) bool {
	return lo.ContainsBy(c.AllowedRepositories, func(repository string) bool {
		regexp, err := regexp.Compile(repository)
		if err != nil {
			return false
		}

		return regexp.MatchString(repository)
	})
}

// GetRunnerName renders the runner name from the template.
func (c *JobLabelConfig) GetRunnerName(runnerID string) (string, error) {
	templ, err := template.New("runner_name").Parse(c.RunnerNameTemplate)
	if err != nil {
		return "", fmt.Errorf("invalid runner_name_template: %w", err)
	}

	var buf strings.Builder
	err = templ.Execute(&buf, map[string]interface{}{
		"ID": runnerID,
	})
	if err != nil {
		return "", fmt.Errorf("failed to execute runner_name_template: %w", err)
	}

	return buf.String(), nil
}

// GetRunnerLabels returns the labels for the runner. It includes the fireactions label and the job label.
func (c *JobLabelConfig) GetRunnerLabels() []string {
	labels := []string{}
	labels = append(labels, c.Name)
	labels = append(labels, c.RunnerLabels...)

	return labels
}

// MetricsConfig is the configuration for the metrics server.
type MetricsConfig struct {
	Enabled bool   `mapstructure:"enabled"`
	Address string `mapstructure:"address"`
}

// Validate validates the configuration.
func (c *MetricsConfig) Validate() error {
	var errs error

	if c.Address == "" {
		errs = multierror.Append(errs, fmt.Errorf("address is required"))
	}

	return errs
}

// HistoryConfig is the configuration for the garbage collection various history data.
type HistoryConfig struct {
	MaxWorkflowRunAge time.Duration `mapstructure:"max_workflow_run_age"`
}

// Validate validates the configuration.
func (c *HistoryConfig) Validate() error {
	var errs error

	if c.MaxWorkflowRunAge == 0 {
		errs = multierror.Append(errs, fmt.Errorf("max_workflow_run_age is required"))
	}

	return errs
}
